Add requirement for braces around most props

Xavientois opened the pull request:
#### Description

<!-- Please include a summary of the change. -->

This PR enforces the usage of braces around property values other than literals (in order to permit normal HTML syntax).
I deviated slightly from the suggestion in [this comment](https://github.com/yewstack/yew/issues/1636#issuecomment-783913553) in order to make the syntax more consistent and similar to JSX). This should open the door to working on [this issue](https://github.com/yewstack/yew/issues/1636).

This is a breaking change. (Might be considered for addition to Component v2?)

All docs and tests have been updated to reflect the new syntax.

Fixes #1775  <!-- replace with issue number or remove if not applicable -->
Fixes #1938 

#### Checklist

<!-- For further details, please read CONTRIBUTING.md -->

- [x] I have run `cargo make pr-flow`
- [X] I have reviewed my own code
- [x] I have added tests
  <!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
  <!-- If this is a feature, my tests prove that the feature works -->

Xavientois commented on packages/yew-macro/src/props/prop.rs:
```diff
@@ -82,6 +83,16 @@ impl Parse for Prop {
     }
 }
 
+fn validate_prop(expr: &Expr) -> syn::Result<()> {
+    match expr {
+        Expr::Block(_) | Expr::Lit(_) => Ok(()),
+        _ => Err(syn::Error::new_spanned(
+            expr,
+            format!("property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {{ {} }}", expr.to_token_stream()),
+        )),
+    }
+}
+
```
This is the actual code corresponding to the change (just so that it its easier to find in this PR)

teymour-aldridge commented on packages/yew-macro/src/props/prop.rs:
```diff
@@ -82,6 +83,16 @@ impl Parse for Prop {
+        Expr::Block(_) | Expr::Lit(_) => Ok(()),
+        _ => Err(syn::Error::new_spanned(
+            expr,
+            format!("property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {{ {} }}", expr.to_token_stream()),
```
nit

```suggestion
            format!("the property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {{ {} }}", expr.to_token_stream()),
```

Xavientois commented on packages/yew/src/virtual_dom/vtag.rs:
```diff
@@ -940,40 +940,14 @@ mod tests {
         elem.detach(&parent);
         assert!(node_ref.get().is_none());
     }
-
```
This PR introduces a breaking change.
Should I update the code to allow this test to pass with braces inserted?

hamza1311 replied:
I think breaking changes are fine. Yew makes no version guarantees right now (v0.x) and making breaking changes where needed is a good idea.

Xavientois commented:
In all of the merge conflicts, the issue appears to be that string literals were turned into owned `String`s and moved data was cloned. Is passing props by reference no longer valid?

hamza1311 commented:
This PR needs master to be rebased onto/merged into it.

Xavientois commented:
> This PR needs master to be rebased onto/merged into it.

Yup, didn't get a chance yesterday, but I'm gonna do it this morning

hamza1311 commented:
Great. Also, 

> In all of the merge conflicts, the issue appears to be that string literals were turned into owned `String`s and moved data was cloned. Is passing props by reference no longer valid?

It never was. Yew stopped implicitly cloning the value and instead of using String which required heap allocation for everything, it now uses `Cow<'static, str>` which only requires heap allocation for non-literal strings

Xavientois commented:
Okay, so now that I have fixed the merge conflicts, when I run `cargo make pr-flow`, everything passes until I get this:
```
error: no library targets found in package `changelog`
[cargo-make][3] ERROR - Error while executing command, exit code: 101
[cargo-make][3] WARN - Build Failed.
[cargo-make][2] ERROR - Error while executing command, exit code: 1
[cargo-make][2] WARN - Build Failed.
[cargo-make][1] ERROR - Error while executing command, exit code: 1
[cargo-make][1] WARN - Build Failed.
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

mc1098 commented on packages/yew-macro/src/props/prop.rs:
```diff
@@ -54,6 +54,30 @@ impl Parse for Prop {
+        Expr::Lit(_) | Expr::Block(_) => Ok(expr),
+        _ => Err(syn::Error::new_spanned(
+                &expr,
+                format!("the property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {{ {} }}", expr.to_token_stream()),
```
```suggestion
                format!("the property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {{{}}}", expr.to_token_stream()),
```
This might be a bit too pedantic, however, this follows the style of all the changes made in this PR.
Following this message might cause users to think that spaces in the braces is the 'correct' format while all the docs are using snippets without spaces.

mc1098 reviewed (commented):
There are changes to i18n documentation for versions 0.17.3 & 0.18.0 (for each translation), which I think is a mistake as these versions won't be effected by the code changes. 

I have only found three instances of missed brackets: 
- [yewtill/src/effect.rs (L60-L63)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/packages/yewtil/src/effect.rs#L60-L63)
- [website/docs/concepts/components/children.md (L171-L173)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/website/docs/concepts/components/children.md#L171-L173)
- [website/docs/concepts/html/classes.md (L106-L110)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/website/docs/concepts/html/classes.md#L106-L110)

Apart from these little bits - this PR is looking good to me :D

Xavientois commented:
> There are changes to i18n documentation for versions 0.17.3 & 0.18.0 (for each translation), which I think is a mistake as these versions won't be effected by the code changes.
> 
> I have only found three instances of missed brackets:
> 
>     * [yewtill/src/effect.rs (L60-L63)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/packages/yewtil/src/effect.rs#L60-L63)
> 
>     * [website/docs/concepts/components/children.md (L171-L173)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/website/docs/concepts/components/children.md#L171-L173)
> 
>     * [website/docs/concepts/html/classes.md (L106-L110)](https://github.com/Xavientois/yew/blob/add-requirement-for-braces-around-props/website/docs/concepts/html/classes.md#L106-L110)


Thanks for catching these! I've reverted all the docs for non-current versions and fixed the missing changes.

mc1098 commented on packages/yew-macro/tests/html_macro/list-fail.stderr:
```diff
@@ -46,11 +46,11 @@ error: expected an expression following this equals sign
 
-error: this closing tag has no corresponding opening tag
-  --> $DIR/list-fail.rs:20:36
+error: the property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {"key" . to_string()}
```
'{"key" . to_string()}' 
The spacing caused by the Display impl of TokenStream is a niggle.
I honestly don't know of a good solution to this - I did see a crate to 'faithfully' display a TokenStream but seems way overkill for this one use case (and I don't even know if it works :P) 
Not sure if this is me just being super picky or whether we leave out the example so that the last sentence would read: 'Consider adding braces around your expression`. 
sorry @Xavientois should have noticed this before the other suggestion about spacing 😞

Xavientois replied:
Yeah, I was trying to figure out how to fix that issue.
Is [this](https://crates.io/crates/proc-macro-faithful-display) the crate you're referring to? The implementation is relatively small and contained to a single file, so for now, I can leave the example out, and then once this PR merges, I can see if there is a straightforward way to get it to display correctly

mc1098 replied:
Yep, that's the crate I found. 
I think without the example the message is clear enough, and when the expression is larger the example probably adds noise.

I've just been playing with some of the examples with this PR and here is just a snippet from the boids example but instead of using the macro I just manually expanded it to see what a larger expression would look like:
```rust
html! {
    <div class="settings">
        <Slider label="Number of Boids"
            min=1.0 max=600.0
            onchange=link.callback(move |value| {
                let mut settings = settings.clone();
                settings.boids = value as usize;
                Msg::ChangeSettings(settings)
            })
            value={settings.boids as f64}
        />
    // rest of the html block 
```
This returns the following error:
```bash
error: the property value must be either a literal or enclosed in braces. Consider adding braces around your expression like this: {link .
callback(move | value |
         {
             let mut settings = settings . clone() ; settings . boids = value
             as usize ; Msg :: ChangeSettings(settings)
         })}
   --> examples/boids/src/main.rs:122:30
    |
122 |                       onchange=link.callback(move |value| {
    |  ______________________________^
123 | |                         let mut settings = settings.clone();
124 | |                         settings.boids = value as usize;
125 | |                         Msg::ChangeSettings(settings)
126 | |                     })
    | |______________________^
```

Without adding the example:
```bash
error: the property value must be either a literal or enclosed in braces. Consider adding braces around your expression.
   --> examples/boids/src/main.rs:122:30
    |
122 |                       onchange=link.callback(move |value| {
    |  ______________________________^
123 | |                         let mut settings = settings.clone();
124 | |                         settings.boids = value as usize;
125 | |                         Msg::ChangeSettings(settings)
126 | |                     })
    | |______________________^
```
I think this is much nicer now that I've laid them out here. The message is clear about the error and the span tells me exactly where I should add those braces I've been told to consider! :D

Xavientois replied:
Done

siku2 commented:
Sorry for the merge conflicts :sweat_smile:

Xavientois commented:
> Sorry for the merge conflicts 😅

@siku2 do you think it's worth fixing these conflicts now? I'd like to wait until maintainers are able to review and merge it so that I don't need to be regularly fixing merge conflicts until then

siku2 commented:
> I'd like to wait until maintainers are able to review and merge it so that I don't need to be regularly fixing merge conflicts until then

Once all the review comments are addressed this should be ready to be merged as far as I'm concerned.

Xavientois commented on examples/nested_list/src/list.rs:
```diff
@@ -102,8 +102,8 @@ impl Component for List {
         html! {
-            <div class="list-container" onmouseout=onmouseout onmouseover=onmouseover>
-                <div class=classes!("list", inactive)>
+            <div class="list-container" onmouseout={onmouseout} onmouseover={onmouseover}>
```
After this PR is merged, the next thing I will work on is collapsing this down to:
```suggestion
            <div class="list-container" {onmouseout} {onmouseover}>
```

Xavientois commented:
> Once all the review comments are addressed this should be ready to be merged as far as I'm concerned.

@siku2 Done! Fixed conflicts, tested, and reviewed twice!
